QB4W SIMD Packing for neondot

mcr229 opened the pull request:
### Summary
We add scalar, neondot, and neoni8mm for packing 4-bit blockwise quantized kernels for GEMM. These kernels are specifically only for 16c4 and 16c8 GEMMs. These kernels help provide significant uplift to packing performance and subsequently model load time performance. We productionize only neondot and neoni8mm kernels despite also writing scalar kernels. We add tests and benchmarks for blockwise packing routines. A key note about these kernels is that they do a one shot initialization of bias and scales as well, unlike existing reference kernels which just use the extra bytes field to leave wholes in the packed weights that are later filled by microparams-init functions, the new ukernels packs the weights with the bias and scales.

### Impact
All performance metrics gathered are on SAMSUNG S24 device. Build modes and flags are all built through the scripts provided in scripts/build-android-arm64.sh.

Average of 10 Runs:
```
qb4_packw_x16c4_goi__reference/llm/B:1/M:128/N:4096/K:1024/real_time_mean                                4905074 ns      4887502 ns           10 bytes=1.77032G/s cpufreq=2.3808G elements=855.096M/s
qb4_packw_x16c8_goi__reference/llm/B:1/M:128/N:4096/K:1024/real_time_mean                                4320939 ns      4304560 ns           10 bytes=2.00964G/s cpufreq=2.3808G elements=970.693M/s
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c8__scalar_llm/B:1/M:128/N:4096/K:1024/real_time_mean        1070324 ns      1066204 ns           10 bytes=8.11299G/s cpufreq=2.3808G elements=3.91873G/s
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c4__scalar_llm/B:1/M:128/N:4096/K:1024/real_time_mean        1223852 ns      1219007 ns           10 bytes=7.09524G/s cpufreq=2.3808G elements=3.42713G/s
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c4__neondot_llm/B:1/M:128/N:4096/K:1024/real_time_mean        342984 ns       341661 ns           10 bytes=25.3176G/s cpufreq=2.3808G elements=12.2289G/s
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c8__neoni8mm_llm/B:1/M:128/N:4096/K:1024/real_time_mean       288537 ns       287416 ns           10 bytes=30.0951G/s cpufreq=2.3808G elements=14.5365G/s
```

In general we see 4x improvement from reference --> scalar and ~15x improvement from reference --> SIMD. Binary size impact of these changes:
```
   0.2%  11.5Ki   1.3%  11.4Ki    xnn_qb4_packw_gemm_goi_ukernel_x16c8__scalar
   0.1%  7.46Ki   0.9%  7.36Ki    xnn_qb4_packw_gemm_goi_ukernel_x16c4__scalar
   0.1%  5.42Ki   0.5%  4.09Ki    xnn_qb4_packw_gemm_goi_ukernel_x16c4__neondot
   0.1%  3.90Ki   0.4%  3.80Ki    xnn_qb4_packw_gemm_goi_ukernel_x16c8__neoni8mm 
```
Scalar kernels significantly contribute to binary size due to the excessive loop unrolling, neondot and neoni8mm kernels are better.

### Additional Experiment
In the commit titled Reduce Binary Size Impact, we experiment with removing the tail handling for when NC % NR != 0:
```
if XNN_UNLIKELY(n != 0){
```

Instead we move this check into the the loop body:
```
for (;n > 0; n -= 16) {
```
and just clamp the weights and scales if n < 16. This surprisingly didn't have a huge performance impact:

```
qb4_packw_x16c4_goi__reference/llm/B:1/M:128/N:4096/K:1024/real_time_mean                                4911195 ns      4893393 ns           10 bytes=1.76811G/s cpufreq=2.3808G elements=854.029M/s
qb4_packw_x16c8_goi__reference/llm/B:1/M:128/N:4096/K:1024/real_time_mean                                4324329 ns      4308086 ns           10 bytes=2.00806G/s cpufreq=2.3808G elements=969.932M/s
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c8__scalar_llm/B:1/M:128/N:4096/K:1024/real_time_mean        1067370 ns      1063305 ns           10 bytes=8.13545G/s cpufreq=2.3808G elements=3.92957G/s
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c4__scalar_llm/B:1/M:128/N:4096/K:1024/real_time_mean        1185798 ns      1181115 ns           10 bytes=7.32294G/s cpufreq=2.3808G elements=3.53712G/s
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c4__neondot_llm/B:1/M:128/N:4096/K:1024/real_time_mean        342087 ns       340746 ns           10 bytes=25.3839G/s cpufreq=2.3808G elements=12.2609G/s
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c8__neoni8mm_llm/B:1/M:128/N:4096/K:1024/real_time_mean       281989 ns       280882 ns           10 bytes=30.7939G/s cpufreq=2.3808G elements=14.874G/s
```

But allowed us to save binary size by 50%:
```
   0.1%  6.19Ki   0.7%  6.09Ki    xnn_qb4_packw_gemm_goi_ukernel_x16c8__scalar
   0.1%  4.06Ki   0.5%  3.96Ki    xnn_qb4_packw_gemm_goi_ukernel_x16c4__scalar
   0.0%  2.51Ki   0.3%  2.42Ki    xnn_qb4_packw_gemm_goi_ukernel_x16c4__neondot
   0.0%  2.38Ki   0.3%  2.28Ki    xnn_qb4_packw_gemm_goi_ukernel_x16c8__neoni8mm
```

Since neondot and neoni8mm are the only productionized kernels, I think the binary size impact seems tolerable here. This branch contains the changes introduced to reduce binary size.

### Further Consierations
I believe the microkernel generation templates could be further generalized for the QC4 case, but due to some complexity of special handling for QB4, I've chosen to leave that for later

Would also like to follow up with multithreading these kernels. I've experimented and found that there can be some significant improvements in load time with this.

### Edits

Updated the branch for neondot kernel to use `vdot` insted`vpadd` we saw the following improvement:
```
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c4__neondot_llm/B:1/M:128/N:4096/K:1024/real_time_mean        307581 ns       306382 ns           10 bytes=28.2317G/s cpufreq=2.3808G elements=13.6364G/s
```
And
```
   0.3%  2.34Ki   0.3%  2.34Ki    _xnn_qb4_packw_gemm_goi_ukernel_x16c4__neondot
```

mcr229 commented:
@alankelly @fbarchard  for reviews

digantdesai commented on src/qb4-packw/gen/qb4-packw-x16c4-gemm-goi-neondot.c:
```diff
@@ -0,0 +1,405 @@
+  const uint8x16_t v1357 = vreinterpretq_u8_u32(vuzp2q_u32(vreinterpretq_u32_u8(v0123), vreinterpretq_u32_u8(v4567)));
+  const uint8x16_t vl1357 = vshlq_n_u8(v1357, 4);
+  const uint8x16_t v01234567 = vorrq_u8(v0246, vl1357);
+  return veorq_u8(v01234567, veor_mask);
```
didn't qc4w move to int8?

digantdesai commented on src/qb4-packw/gen/qb4-packw-x16c4-gemm-goi-neondot.c:
```diff
@@ -0,0 +1,405 @@
+{
+  const uint8x16_t vl = vshrq_n_u8(v, 4);  // isolate lower int 4
+  const uint8x16_t vh = vandq_u8(v, vmask);  // isolate upper int 4
+  const uint32x4_t vh_acc = vpaddlq_u16(vpaddlq_u8(vh));
+  const uint32x4_t vl_acc = vpaddlq_u16(vpaddlq_u8(vl));
```
use dotprod with zero here

mcr229 replied:
ahh oops, i had this in another branch but it failed to get in here. Let me update

digantdesai commented on src/qb4-packw/gen/qb4-packw-x16c4-gemm-goi-neondot.c:
```diff
@@ -0,0 +1,405 @@
+      out += 16 * sizeof(float);
+
+      // KC/2 bytes is KC Nibbles
+      const uint8_t* w1 = w0 + (kc >> 1);
```
hoist `kc >> 1` in a variable

digantdesai commented on src/qb4-packw/gen/qb4-packw-x16c4-gemm-goi-neondot.c:
```diff
@@ -0,0 +1,405 @@
+          uint32x4_t wEx0123 = vld1q_u32((uint32_t*) w14);
+          uint32x4_t wFx0123 = vld1q_u32((uint32_t*) w15);
+
+          uint32x4_t v01_02 = vtrn1q_u32(w0x0123, w1x0123);
```
are we spilling in this block? if yes, interleave transpose with load?

digantdesai commented on src/qb4-packw/gen/qb4-packw-x16c4-gemm-goi-neondot.c:
```diff
@@ -0,0 +1,405 @@
+          uint32x4_t vCDEF_1 = vcombine_u32(vget_low_u32(vCD_13), vget_low_u32(vEF_13));
+          uint32x4_t vCDEF_3 = vcombine_u32(vget_high_u32(vCD_13), vget_high_u32(vEF_13));
+
+          v0123_0 = xnn_packed2planar(&ksum0123, v0123_0, vmask, veor_mask, neg_zp);
```
nit `veor_mask` - why does it have to be an arg?

mcr229 replied:
not sure what the alternative is for this? Do you mean to do the or outside of packed2planar?

digantdesai commented on src/qb4-packw/gen/qb4-packw-x16c4-gemm-goi-neondot.c:
```diff
@@ -0,0 +1,405 @@
+          uint32x4_t vCDEF_1 = vcombine_u32(vget_low_u32(vCD_13), vget_low_u32(vEF_13));
+          uint32x4_t vCDEF_3 = vcombine_u32(vget_high_u32(vCD_13), vget_high_u32(vEF_13));
+
+          v0123_0 = xnn_packed2planar(&ksum0123, v0123_0, vmask, veor_mask, neg_zp);
```
Also did you consider using `tbl` instruction instead of this algo for planner packing?

digantdesai commented on src/qb4-packw/gen/qb4-packw-x16c4-gemm-goi-neondot.c:
```diff
@@ -0,0 +1,405 @@
+          out += 256;
+        }
+
+        uint16x8_t bf_scales01234567 = {
```
does this result in 16 loads? Is it temporary until the transposed_scales flag?

mcr229 replied:
yea it is done in place of transposed scales flag. Transposed scales can give ~10% extra uplift. But it's a lot more working wiring things up, so i just decided to do this as short term. -10% perf isn't that bad compared to how much better compared to scalar both in binary size and perf. We can think of transposed scales later, but the one concern about that is what the best way to multi thread that would be, since scales is read only it wouldn't be that bad, but passing scale stride could be a bit nasty later on.

digantdesai commented on src/qb4-packw/kr-neoni8mm.c.in:
```diff
@@ -0,0 +1,192 @@
+{
+  const uint8x16_t vl = vshrq_n_u8(v, 4);  // isolate lower int 4
+  const uint8x16_t vh = vandq_u8(v, vmask);  // isolate upper int 4
+  *vacc = vreinterpretq_s32_u32(vmmlaq_u32(vreinterpretq_u32_s32(*vacc), vones_zero_mask, vl));
```
too much to have this template just for these instructions? can we drop it and just have dot prod one? What is the perf delta for the same output tile size between dot and i8mm?

mcr229 replied:
so perhaps this is bad naming of the template, but I think there is a good amount of difference is also due to kr=8 vs kr=4. Specifically the rearranging of the planes and blocks before we read xnn_packed2planar. I take advantage of the fact different KR is also different instructions to have different templates.

digantdesai reviewed (commented):
Looks great @mcr229 . Left some comments, I will let Alan, Frank or Dylan to stamp.

mcr229 commented on bench/bgemm.h:
```diff
@@ -100,3 +102,14 @@ static void QD8AttentionBgemmArguments(benchmark::internal::Benchmark* b) {
+  b->Args({1, 128, 16, 1024});
+  b->Args({1, 128, 128, 1024});
+  b->Args({1, 128, 4096, 1024});
+  b->Args({1, 128, 11008, 4096});
+  b->Args({1, 128, 32000, 4096});
```
These big ones actually OOM my S24, it works on my mac though. I might remove them since they can't run on my phone.

mcr229 commented:
Force push changes: 
- resolved some nits in the ukernels
- removed OOM-ing benchmarks
- Use vdot instead of padd in neondot

mcr229 commented:
@alankelly @dsharlet @fbarchard

fbarchard commented on bench/packw-benchmark.h:
```diff
@@ -910,3 +978,44 @@ BENCHMARK_BGEMM(x32_packw_x8_gio__reference)
 BENCHMARK_BGEMM(x32_packw_x32_gio__reference)
 BENCHMARK_BGEMM(x32_packw_x64_gio__reference)
+
+static void qb4_packw_goi__reference(
```
consider gio reference, which is normally easy, but c4/c8 make gio harder than goi in some ways.
a scalar gio would be a good step before doing neon.

mcr229 replied:
yea i considered doing gio, but it seemed a lot more difficult for my ROI (i don't think i have any use case with transposed weights ATM). That might come back later for me to eventually finish, but ok with doing goi for now.

fbarchard commented on src/qb4-packw/kr-neoni8mm.c.in:
```diff
@@ -0,0 +1,193 @@
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+$ABC = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
```
bit avoid $ABC moving forward.  replace
${ABC[N:N+4]}
with
${N//4}
this is because future cpus have very large vectors, we have 80 column code style, and sve & rvv dont have known vector sizes.
If you have complicated lanes dont worry about it, but it ${N} or ${N//4} works its an easy cleanup

fbarchard commented on src/qb4-packw/kr-neoni8mm.c.in:
```diff
@@ -0,0 +1,193 @@
+{
+  const int8x16_t vl = vshrq_n_s8(v, 4);  // isolate lower int 4
+  const int8x16_t vh = vandq_s8(v, vmask);  // isolate upper int 4
+  *vacc = vmmlaq_s32(*vacc, vones_zero_mask, vl);
```
so this kernel is i8mm because you sum up KR=8 using vmmlaq?
it is worth it?  You could do sdot and paired add outside loop

We've done neon c8 kernels before because i8mm is not really a win for m=1

fbarchard reviewed (changes requested):
overall good.
I've observed the scalar packw code size on hexagon where NR=256 is a bit excessive unrolling, as well as a bottleneck, but I'll deal with that by using SIMD, not scalar

mcr229 commented:
New Changes:
- remove $ABC from packw templates
- move Neoni8mm kernel to be neondot friendly.
- clang-format

@fbarchard for another review

fbarchard commented on src/qb4-packw/gen/qb4-packw-x16c4-gemm-goi-neondot.c:
```diff
@@ -0,0 +1,405 @@
+// clang-format off
+// Auto-generated file. Do not edit!
+//   Template: src/qb4-packw/kr4-neondot.c.in
```
nit: file name convention is c4-neondot.c.in

fbarchard commented on src/qb4-packw/gen/qb4-packw-x16c4-gemm-goi-neondot.c:
```diff
@@ -0,0 +1,405 @@
+  const int32x4_t neg_zp,
+  const int8x16_t vones)
+{
+  const uint8x16_t vl = vshrq_n_u8(v, 4);  // isolate lower int 4
```
vshlq_n_u8 ?

mcr229 replied:
we accumulate the shl, because we want to save the right shift that's done in the inner block loop of the gemm.

qc4 has a right shift here:
https://github.com/google/XNNPACK/blob/master/src/qd8-f32-qc4w-gemm/gen/qd8-f32-qc4w-gemm-1x1-minmax-scalar.c#L62

but for blockwise it would be more expensive because we would have to do this shift after every block:
https://github.com/google/XNNPACK/blob/7e2e76bf01eba1886ef908cedb6324f5e09dfc0c/src/qd8-f32-qb4w-gemm/gen/qd8-f32-qb4w-gemm-2x4-minmax-scalar.c#L116

the nibbles are still left shifted in our gemm loop, so we instead scale our kernel_scales by 1/16:
https://github.com/google/XNNPACK/blob/7e2e76bf01eba1886ef908cedb6324f5e09dfc0c/src/microparams-init.c#L448


I do still have to account for kernel_zero_point = 0 though, which i can add

fbarchard commented on src/reference/packing.cc:
```diff
@@ -1854,12 +1857,16 @@ void xnn_pack_qb4_weights_and_biases(
         /*bl=*/block_size,
         /*kernel=*/(const uint8_t*)weights,
-        /*bias=*/nullptr,
+        /*bias=*/has_fast_packing_ukernel ? (const int32_t*)accumulator_init : nullptr,
```
this might be worth a comment?

mcr229 replied:
sounds good will do

mcr229 commented:
Updates:
- Updated kernels to support both signed and unsigned (zero_point = 0 & zero_point = 8)
- Added comment for initializing bias in packing kernel
- renamed templates to c4-neondot.in and c8-neondot.in

Performance Updates:
Updating to support signed and unsigned seems to have a large effect on the scalar kernels. Performance regresses ~10%, and for some reason the scalar c4 kernel's binary size almost triples from 3.96 --> ~10(c8 scalar goes down). I couldn't figure out why. But since scalar kernels aren't productionized here I didn't let it go(i tried looking into it for some time but couldn't figure it out). SIMD Kernels don't have much change:

Binary Size
```
   0.1%  10.4Ki   1.2%  10.3Ki    xnn_qb4_packw_gemm_goi_ukernel_x16c4__scalar
   0.1%  5.71Ki   0.6%  5.61Ki    xnn_qb4_packw_gemm_goi_ukernel_x16c8__scalar
   0.1%  3.63Ki   0.3%  2.29Ki    xnn_qb4_packw_gemm_goi_ukernel_x16c4__neondot
   0.0%  2.42Ki   0.3%  2.32Ki    xnn_qb4_packw_gemm_goi_ukernel_x16c8__neondot
```

Perf
```
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c8__scalar_llm/B:1/M:128/N:4096/K:1024/real_time_mean       1206842 ns      1202209 ns           10 bytes=7.19524G/s cpufreq=2.3808G elements=3.47544G/s
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c4__scalar_llm/B:1/M:128/N:4096/K:1024/real_time_mean       1387612 ns      1382595 ns           10 bytes=6.25789G/s cpufreq=2.3808G elements=3.02268G/s
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c4__neondot_llm/B:1/M:128/N:4096/K:1024/real_time_mean       291198 ns       290106 ns           10 bytes=29.82G/s cpufreq=2.3808G elements=14.4036G/s
qb4_packw/xnn_qb4_packw_gemm_goi_ukernel_x16c8__neondot_llm/B:1/M:128/N:4096/K:1024/real_time_mean       280933 ns       279850 ns           10 bytes=30.9096G/s cpufreq=2.3808G elements=14.9299G/s
```

cc. @fbarchard

mcr229 commented:
@fbarchard @dsharletg @alankelly 

slight nudge for hopefully a final review :)

mcr229 commented:
Rebasing

dsharlet commented on test/CMakeLists.txt:
```diff
@@ -125,6 +125,7 @@ SET(MICROKERNEL_UNIT_TESTS
     f32-vscaleextexp
     indirection
     packing
+    qb4-packw
```
Add to bazel BUILD too

dsharlet reviewed (approved):
@fbarchard you specifically requested changes in a previous review, can you please take a look?

mcr229 commented:
resolved remaining build failures

dsharlet commented:
This is broken for us with many errors of the form:
```
src/qb4-packw/gen/qb4-packw-x16c4-gemm-goi-scalar.c:23:38: error: cast from 'const unsigned char *' to 'unsigned int *' drops const qualifier [-Werror,-Wcast-qual]
   23 |   const uint32_t s_v0 = (((uint32_t*)weights)[0] ^ vkernel_zero_point);
```

mcr229 commented:
> This is broken for us with many errors of the form:
> 
> ```
> src/qb4-packw/gen/qb4-packw-x16c4-gemm-goi-scalar.c:23:38: error: cast from 'const unsigned char *' to 'unsigned int *' drops const qualifier [-Werror,-Wcast-qual]
>    23 |   const uint32_t s_v0 = (((uint32_t*)weights)[0] ^ vkernel_zero_point);
> ```

let me fix that, thanks for the catch.

dsharlet commented:
This is still not building:
```
XNNPACK/src/qb4-packw/gen/qb4-packw-x16c8-gemm-goi-aarch64-neondot.c:71:19: error: unused variable 'veor_mask' [-Werror,-Wunused-variable]
   71 |   const int8x16_t veor_mask = vmovq_n_s8(UINT8_C(0x88));
      |                   ^~~~~~~~~
XNNPACK/src/qb4-packw/gen/qb4-packw-x16c8-gemm-goi-aarch64-neondot.c:72:19: error: unused variable 'neg_zp' [-Werror,-Wunused-variable]
   72 |   const int32x4_t neg_zp = vmovq_n_s32(-64);
      |                   ^~~~~~
XNNPACK/src/qb4-packw/gen/qb4-packw-x16c8-gemm-goi-aarch64-neondot.c:74:18: error: unused variable 'vzeros' [-Werror,-Wunused-variable]
   74 |   const int8x8_t vzeros = vmov_n_s8(0);
```

dsharlet commented on src/qb4-packw/gen/qb4-packw-x16c8-gemm-goi-scalar.c:
```diff
@@ -0,0 +1,381 @@
+  int8_t* packed_weights,
+  size_t extra_bytes_bl,
+  size_t extra_bytes_n,
+  const void* params)
```
This kernel either needs to be attributed `XNN_OOB_READS`, or it has a bug that causes it to read out of bounds:
```
[ RUN      ] qb4_packw/XnnTestQB4.nc_gt_nr/xnn_qb4_packw_gemm_goi_ukernel_x16c8__scalar
=================================================================
==10==ERROR: AddressSanitizer: use-after-poison on address 0x59b00154 at pc 0x006d9f94 bp 0x407ff040 sp 0x407ff03c
READ of size 4 at 0x59b00154 thread T0
```
(call stack is useless unfortunately)

mcr229 replied:
i'll figure this one out thanks!

mcr229 replied:
the issue was an out of bounds read of the bias, since our packing kernels also do a one shot bias fill for packing. I marked the packing kernels as XNN_OOB_READ.
